File: src/main/java/org/distril/beengine/world/World.java
package org.distril.beengine.world;

import com.nukkitx.math.vector.Vector3f;
import com.nukkitx.math.vector.Vector3i;
import com.nukkitx.protocol.bedrock.packet.UpdateBlockPacket;
import lombok.Getter;
import lombok.extern.log4j.Log4j2;
import org.distril.beengine.Tickable;
import org.distril.beengine.entity.Entity;
import org.distril.beengine.material.Material;
import org.distril.beengine.material.block.Block;
import org.distril.beengine.material.item.Item;
import org.distril.beengine.player.Player;
import org.distril.beengine.server.Server;
import org.distril.beengine.util.Direction;
import org.distril.beengine.world.chunk.Chunk;
import org.distril.beengine.world.chunk.ChunkManager;
import org.distril.beengine.world.generator.Generator;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Getter
@Log4j2
public class World extends Tickable {

	private static final int MAX_Y = 256;
	private static final int MIN_Y = 0;

	private final Path path;

	private final String worldName;
	private final Dimension dimension;
	private final Generator generator;

	private final ChunkManager chunkManager = new ChunkManager(this);
	private final Map<Long, Entity> entities = new ConcurrentHashMap<>();

	public World(String worldName, Dimension dimension, Generator generator) {
		super(worldName + " - World");
		this.path = Path.of("./worlds/" + worldName);

		try {
			Files.createDirectories(this.path);
		} catch (IOException ignored) {/**/}

		this.worldName = worldName;
		this.dimension = dimension;
		this.generator = generator;

		this.start();
	}

	@Override
	protected void onUpdate(long currentTick) {
		this.entities.values()
				.stream()
				.filter(Objects::nonNull)
				.forEach(entity -> entity.onUpdate(currentTick));
	}

	public void addEntity(Entity entity) {
		this.addEntity(entity, false);
	}

	public void addEntity(Entity entity, boolean addInChunk) {
		this.entities.put(entity.getId(), entity);

		if (addInChunk) {
			var location = entity.getLocation();

			location.getChunk().addEntity(entity);
		}
	}

	public void removeEntity(Entity entity) {
		this.entities.remove(entity.getId());
	}

	public void removeEntity(Entity entity, boolean removeInChunk) {
		this.entities.remove(entity.getId());

		if (removeInChunk) {
			var location = entity.getLocation();

			location.getChunk().removeEntity(entity);
		}
	}

	public Chunk getLoadedChunk(int x, int z) {
		return this.chunkManager.getLoadedChunk(x, z);
	}

	public Chunk getLoadedChunk(long key) {
		return this.chunkManager.getLoadedChunk(key);
	}

	public Chunk getChunk(Vector3i position) {
		return this.chunkManager.getChunk(position.getX() >> 4, position.getZ() >> 4);
	}

	public Chunk getChunk(int x, int z) {
		return this.chunkManager.getChunk(x, z);
	}

	public Chunk getChunk(long key) {
		return this.chunkManager.getChunk(key);
	}

	public Block getLoadedBlock(Vector3i position) {
		return this.getLoadedBlock(position.getX(), position.getY(), position.getZ());
	}

	public Block getLoadedBlock(int x, int y, int z) {
		return this.getLoadedBlock(x, y, z, 0);
	}

	public Block getLoadedBlock(int x, int y, int z, int layer) {
		int chunkX = x >> 4;
		int chunkZ = z >> 4;

		var chunk = this.getLoadedChunk(chunkX, chunkZ);

		if (MIN_Y < 0 || y >= MAX_Y) {
			var block = Material.AIR.getBlock();
			block.setPosition(Vector3i.from(x, y, z));
			return block;
		}

		if (chunk == null) {
			return null;
		}

		var block = chunk.getBlock(x & 0xf, y, z & 0xf, layer);
		block.setPosition(this, Vector3i.from(x, y, z));
		return block;
	}

	public Block getBlock(Vector3i position) {
		return this.getBlock(position.getX(), position.getY(), position.getZ());
	}

	public Block getBlock(int x, int y, int z) {
		return this.getBlock(x, y, z, 0);
	}

	public Block getBlock(int x, int y, int z, int layer) {
		int chunkX = x >> 4;
		int chunkZ = z >> 4;

		var chunk = this.getChunk(chunkX, chunkZ);

		if (MIN_Y < 0 || y >= MAX_Y) {
			var block = Material.AIR.getBlock();
			block.setPosition(Vector3i.from(x, y, z));
			return block;
		}

		if (chunk == null) {
			return null;
		}

		var block = chunk.getBlock(x & 0xf, y, z & 0xf, layer);
		block.setPosition(this, Vector3i.from(x, y, z));
		return block;
	}

	public void setBlock(Vector3i position, int layer, Block block) {
		this.setBlock(position.getX(), position.getY(), position.getZ(), layer, block, true);
	}

	public void setBlock(int x, int y, int z, int layer, Block block) {
		this.setBlock(x, y, z, layer, block, true);
	}

	public void setBlock(int x, int y, int z, int layer, Block block, boolean send) {
		if (y < MIN_Y || y >= MAX_Y) {
			return;
		}

		var chunk = this.getChunk(x >> 4, z >> 4);
		chunk.setBlock(x & 0xF, y, z & 0xF, layer, block);

		if (send) {
			this.sendBlocks(chunk.getPlayers(), Collections.singleton(block), UpdateBlockPacket.FLAG_ALL_PRIORITY);
		}
	}

	public void sendBlocks(Collection<Player> targets, Collection<Block> blocks, Set<UpdateBlockPacket.Flag> flags) {
		blocks.stream().filter(Objects::isNull).forEach(block -> {
			throw new NullPointerException("Null block is update array");
		});

		List<UpdateBlockPacket> packets = new ArrayList<>();

		blocks.forEach(block -> {
			var packet = new UpdateBlockPacket();
			packet.getFlags().addAll(flags);
			packet.setBlockPosition(block.getPosition());
			packet.setRuntimeId(block.getState().getRuntimeId());
			packet.setDataLayer(0);

			packets.add(packet);

			packet.setDataLayer(1);
			packets.add(packet);
		});

		Server.getInstance().broadcastPackets(targets, packets);
	}

	public Item useItemOn(Vector3i blockPosition, Item item, Direction blockFace, Vector3f clickPosition, Player player) {
		return this.useItemOn(blockPosition, item, blockFace, clickPosition, player, true);
	}

	public Item useItemOn(Vector3i blockPosition, Item item, Direction blockFace, Vector3f clickPosition, Player player, boolean playSound) {
		var target = this.getBlock(blockPosition);
		var block = target.getSide(blockFace);

		Vector3i blockPos = block.getPosition();

		if (blockPos.getY() >= MAX_Y || blockPos.getY() < MIN_Y) {
			return null;
		}

		if (target.getMaterial() == Material.AIR) {
			return null;
		}

		this.setBlock(blockPos, 0, target);

		return item;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}

		if (obj == null || this.getClass() != obj.getClass()) {
			return false;
		}

		World that = (World) obj;
		return this.path.equals(that.getPath()) && this.worldName.equals(that.getWorldName());
	}

	@Override
	public int hashCode() {
		return Objects.hash(this.path, this.worldName);
	}
}

File: src/main/kotlin/org/distril/beengine/world/World.kt
package org.distril.beengine.world

import com.nukkitx.math.vector.Vector3i
import com.nukkitx.protocol.bedrock.packet.UpdateBlockPacket
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import org.distril.beengine.Tickable
import org.distril.beengine.entity.Entity
import org.distril.beengine.material.Material
import org.distril.beengine.material.block.Block
import org.distril.beengine.material.item.Item
import org.distril.beengine.player.Player
import org.distril.beengine.server.Server
import org.distril.beengine.util.Direction
import org.distril.beengine.world.chunk.manager.ChunkManager
import org.distril.beengine.world.chunk.processor.PlayerChunkProcessor
import org.distril.beengine.world.chunk.processor.PlayerChunkRequest
import org.distril.beengine.world.generator.Generator
import java.nio.file.Path
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class World(val worldName: String, val dimension: Dimension, val generator: Generator) : Tickable(
	"$worldName-World"
) {

	val path = Path.of("worlds", worldName)
	val chunkManager = ChunkManager(this)
	val playerChunkProcessor = PlayerChunkProcessor(this.worldName)
	val entities: MutableMap<Long, Entity> = ConcurrentHashMap()

	init {
		this.start()
	}

	override suspend fun onUpdate(currentTick: Long): Unit = coroutineScope {
		val tickedItems = mutableListOf<Deferred<Any>>()
		tickedItems.addAll(chunkManager.tick())

		entities.values.forEach {
			tickedItems.add(async { it.onUpdate(currentTick) })
		}

		awaitAll(*tickedItems.toTypedArray())
	}

	fun addPlayerRequest(request: PlayerChunkRequest) {
		this.playerChunkProcessor.addRequest(request)
	}

	fun addEntity(entity: Entity, addInChunk: Boolean = true) {
		this.entities[entity.id] = entity

		if (addInChunk) {
			val location = entity.location
			location.chunk.addEntity(entity)
		}
	}

	fun removeEntity(entity: Entity, removeInChunk: Boolean = true) {
		this.entities.remove(entity.id)

		if (removeInChunk) {
			val location = entity.location
			location.chunk.removeEntity(entity)
		}
	}

	fun getLoadedBlock(position: Vector3i, layer: Int = 0) =
		this.getLoadedBlock(position.x, position.y, position.z, layer)

	fun getLoadedBlock(x: Int, y: Int, z: Int, layer: Int = 0): Block? {
		if (y >= this.dimension.maxY || y < this.dimension.minY) {
			return Material.AIR.getBlock<Block>().apply {
				world = this@World
				position = Vector3i.from(x, y, z)
			}
		}

		val chunkX = x shr 4
		val chunkZ = z shr 4

		val chunk = this.chunkManager.getLoadedChunk(chunkX, chunkZ) ?: return null

		return chunk.getBlock(x and 0xf, y, z and 0xf, layer).apply {
			world = this@World
			position = Vector3i.from(x, y, z)
		}
	}

	fun getBlock(position: Vector3i, layer: Int = 0) = this.getBlock(position.x, position.y, position.z, layer)

	fun getBlock(x: Int, y: Int, z: Int, layer: Int = 0): Block {
		val loadedBlock = this.getLoadedBlock(x, y, z, layer)
		if (loadedBlock != null) return loadedBlock

		val chunkX = x shr 4
		val chunkZ = z shr 4

		val chunk = this.chunkManager.getChunk(chunkX, chunkZ)

		return chunk.getBlock(x and 0xf, y, z and 0xf, layer).apply {
			world = this@World
			position = Vector3i.from(x, y, z)
		}
	}

	fun setBlock(position: Vector3i, block: Block, layer: Int = 0) =
		this.setBlock(position.x, position.y, position.z, block, layer)

	fun setBlock(x: Int, y: Int, z: Int, block: Block, layer: Int = 0, send: Boolean = true) {
		if (y >= this.dimension.maxY || y < this.dimension.minY) return

		val chunk = this.chunkManager.getChunk(x shr 4, z shr 4)
		chunk.setBlock(x and 0xF, y, z and 0xF, block, layer)

		if (send) this.sendBlocks(chunk.players, listOf(block))
	}

	fun sendBlocks(
		targets: Collection<Player>,
		blocks: Collection<Block>,
		flags: Set<UpdateBlockPacket.Flag> = UpdateBlockPacket.FLAG_ALL_PRIORITY
	) {
		val packets = mutableListOf<UpdateBlockPacket>()

		val packet = UpdateBlockPacket()
		packet.flags.addAll(flags)
		blocks.forEach {
			packet.blockPosition = it.position
			packet.runtimeId = it.state.runtimeId
			packet.dataLayer = 0
			packets.add(packet)

			packet.dataLayer = 1
			packets.add(packet)
		}

		Server.broadcastPackets(targets, packets)
	}

	fun getLoadedChunkEntities(chunkX: Int, chunkZ: Int): Collection<Entity> {
		return this.chunkManager.getLoadedChunk(chunkX, chunkZ)?.entities ?: return setOf()
	}

	fun useItemOn(blockPosition: Vector3i, usedItem: Item, clickedBlockFace: Direction, player: Player): Item? {
		val clickedBlock = this.getBlock(blockPosition)
		if (clickedBlock.material === Material.AIR) return null

		val replaceBlock = clickedBlock.getSide(clickedBlockFace)

		val replacePos = replaceBlock.position!!
		if (replacePos.y < this.dimension.maxY && replacePos.y >= this.dimension.minY) {
			usedItem.toBlock<Block>()?.apply {
				world = this@World
				position = replacePos
			}?.run { setBlock(replacePos, block = this) }

			return usedItem
		}

		return null
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is World) return false

		return this.path == other.path
	}

	override fun hashCode() = Objects.hash(this.path)
}
